vector/src/main/java/im/vector/app/core/di/SessionInitializer.kt

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2022 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.core.di
+
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.newSingleThreadContext
+import kotlinx.coroutines.withContext
+import org.matrix.android.sdk.api.auth.AuthenticationService
+import org.matrix.android.sdk.api.session.Session
+import javax.inject.Inject
+
+@OptIn(DelicateCoroutinesApi::class)
+private val INITIALIZER_CONTEXT = newSingleThreadContext("session-initializer")
+
+class SessionInitializer @Inject constructor(
+        private val authenticationService: AuthenticationService,
+) {
+
+    /**
+     * A thread safe way to initialize the last authenticated Session instance.
+     *
+     * @param readCurrentSession expects an in-memory Session to be provided or null if not yet set.
+     * @param initializer callback to allow additional initialization on the Session, such as setting the in-memory Session instance.
+     * @return the initialized Session or null when no authenticated sessions are available.
+     */
+    suspend fun tryInitialize(readCurrentSession: () -> Session?, initializer: (Session) -> Unit): Session? {
+        return withContext(INITIALIZER_CONTEXT) {
```

Comment:
the initialisation is thread safe and will block or wait for any in progress initialisers to complete, this means that ideally all access to `ActiveSessionHolder` should be off the main thread to avoid an possible ANRs, however this is quite a large refactor

the most common flow will be via the `MainActivity` which is off the main thread

Reply:
Maybe safer to use a semaphore instead of a single thread dispatcher?

Reply:
will give it a try :+1:

Reply:
works well! :100: https://github.com/vector-im/element-android/pull/6710/commits/9114630bbaec9020de897d970d7da1e2e739ae8f